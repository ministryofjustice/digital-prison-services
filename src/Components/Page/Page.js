import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { H1 } from '@govuk-react/heading'
import BackLink from '@govuk-react/back-link'
import Error from '../../Error'
import Spinner from '../../Spinner'
import { Container, PageHeader } from './Page.styles'
import { childrenType } from '../../types'
import Breadcrumb from '../Breadcrumb'
import { setApplicationTitle } from '../../redux/actions'
import PrintLink from './elements/PrintLink'

export class Page extends Component {
  componentDidMount() {
    const { title, docTitle, applicationTitle, dispatchApplicationTitle } = this.props
    this.renderTitleString(title, docTitle)

    if (applicationTitle) {
      dispatchApplicationTitle(applicationTitle)
    }
  }

  componentWillUpdate(nextProps) {
    this.renderTitleString(nextProps.title, nextProps.docTitle)
  }

  renderTitleString = (title, docTitle) => {
    const appName = 'Digital Prison Services'
    document.title = title ? `${docTitle || title} - ${appName}` : appName
  }

  render() {
    const { error, loaded, title, children, alwaysRender, showBreadcrumb, homeLink, backLink, showPrint } = this.props

    if (alwaysRender || loaded || error) {
      return (
        <>
          {alwaysRender && !loaded && <Spinner />}
          {showBreadcrumb && <Breadcrumb homeLink={homeLink} />}
          {backLink && <BackLink href={backLink}>Back</BackLink>}
          <Container>
            <PageHeader>
              <H1 size="LARGE">{title}</H1>
              {showPrint && (
                <div>
                  <PrintLink />
                </div>
              )}
            </PageHeader>
            {error && <Error error={error} />}
            {(!error || alwaysRender) && <div className="page-content">{children}</div>}
          </Container>
          {showPrint && <PrintLink bottom />}
        </>
      )
    }

    return <Spinner />
  }
}

Page.propTypes = {
  error: PropTypes.string.isRequired,
  loaded: PropTypes.bool.isRequired,
  title: PropTypes.string,
  docTitle: PropTypes.string,
  children: childrenType,
  alwaysRender: PropTypes.bool,
  showBreadcrumb: PropTypes.bool,
  homeLink: PropTypes.string.isRequired,
  backLink: PropTypes.string,
  applicationTitle: PropTypes.string,
  dispatchApplicationTitle: PropTypes.func,
  showPrint: PropTypes.bool,
}

Page.defaultProps = {
  alwaysRender: false,
  showBreadcrumb: true,
  backLink: '',
  title: undefined,
  docTitle: undefined,
  children: undefined,
  applicationTitle: '',
  dispatchApplicationTitle: () => {},
  showPrint: false,
}

const mapDispatchToProps = dispatch => ({
  dispatchApplicationTitle: title => dispatch(setApplicationTitle(title)),
})

const mapStateToProps = state => ({
  error: state.app.error,
  loaded: state.app.loaded,
  homeLink: state.app.config.notmEndpointUrl,
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Page)
