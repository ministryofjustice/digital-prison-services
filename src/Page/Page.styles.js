import styled from 'react-emotion'
import { SPACING } from '@govuk-react/constants'

export const Breadcrumbs = styled('div')`
  padding-top: ${SPACING.SCALE_3};
`
export const Container = styled('div')`
  padding-top: ${SPACING.SCALE_4};
`
