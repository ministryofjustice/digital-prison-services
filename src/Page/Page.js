import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import Header from '@govuk-react/header'
import Error from '../Error'
import Spinner from '../Spinner'
import { Container, Breadcrumbs } from './Page.styles'
import links from '../links'

export const Page = ({ error, loaded, title, children }) => {
  if (error) return <Error error={error} />

  if (loaded) {
    return (
      <Fragment>
        <Breadcrumbs>
          <a id="back_to_menu_link" title="Back to home link" className="link backlink" href={links.getHomeLink()}>
            <img className="back-triangle" src="/images/BackTriangle.png" alt="" width="6" height="10" /> Back
          </a>
        </Breadcrumbs>
        <Container>
          <Header level={1} size="LARGE">
            {title}
          </Header>
          <div className="page-content">{children}</div>
        </Container>
      </Fragment>
    )
  }

  return <Spinner />
}

Page.propTypes = {
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.shape({ message: PropTypes.string })]).isRequired,
  loaded: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
}

const mapStateToProps = state => ({
  error: state.app.error,
  loaded: state.app.loaded,
})

export default connect(mapStateToProps)(Page)
