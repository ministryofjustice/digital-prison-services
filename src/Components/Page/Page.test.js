import React from 'react'
import { shallow, mount } from 'enzyme'
import { MemoryRouter } from 'react-router'
import { Page } from './Page'

describe('<Page />', () => {
  const props = {
    title: 'Page title',
    error: '',
    loaded: false,
    children: 'Page content',
    homeLink: 'http://home/',
    defaultHome: 'http://notmhome/',
    backLink: 'http://this/',
  }

  describe('when the page is loading', () => {
    const wrapper = shallow(<Page {...props} />)

    it('should display the loading <Spinner /> component', () => {
      expect(wrapper.find('Spinner').exists()).toBe(true)
    })
  })

  describe('when the page has loaded', () => {
    let wrapper

    beforeEach(() => {
      wrapper = mount(
        <MemoryRouter initialEntries={['/random']}>
          <Page {...props} loaded />
        </MemoryRouter>
      )
    })

    it('should update the document title', () => {
      expect(global.window.document.title).toEqual('Page title - Digital Prison Services')
    })

    it('should display the children prop', () => {
      expect(wrapper.find('.page-content').contains(props.children)).toEqual(true)
    })

    it('should display the Breadcrumb component by default', () => {
      expect(wrapper.find('Breadcrumb').exists()).toEqual(true)
    })

    it('should pass the homelink to Breadcrumb component', () => {
      expect(
        wrapper
          .find('Breadcrumb')
          .find({ homeLink: 'http://home/' })
          .exists()
      ).toEqual(true)
    })

    it('should render a back link if prop passed in', () => {
      expect(
        wrapper
          .find('BackLink')
          .find({ href: 'http://this/' })
          .exists()
      ).toEqual(true)
    })

    it('should NOT display the Breadcrumb component when showBreadcrumb is false', () => {
      wrapper = mount(
        <MemoryRouter initialEntries={['/random']}>
          <Page {...props} loaded showBreadcrumb={false} />
        </MemoryRouter>
      )

      expect(wrapper.find('Breadcrumb').exists()).toEqual(false)
    })

    it('should NOT display the back link component when no backLink prop', () => {
      wrapper = mount(
        <MemoryRouter initialEntries={['/random']}>
          <Page {...props} loaded backLink="" />
        </MemoryRouter>
      )

      expect(wrapper.find('BackLink').exists()).toEqual(false)
    })
  })

  describe('when there is an error', () => {
    const error = 'Error message'

    it('should an error message', () => {
      const wrapper = shallow(<Page {...props} loaded error={error} />)

      expect(wrapper.find('Error').exists()).toBe(true)
      expect(wrapper.find('Error').prop('error')).toEqual(error)
    })

    it('should display an error message and page content if alwaysRender is specified', () => {
      const wrapper = shallow(<Page {...props} loaded error={error} alwaysRender />)

      expect(wrapper.find('Error').exists()).toBe(true)
      expect(wrapper.find('Error').prop('error')).toEqual(error)
      expect(wrapper.find('.page-content').contains(props.children)).toEqual(true)
    })

    it('should display spinner and contents', () => {
      const wrapper = shallow(<Page {...props} alwaysRender loaded={false} />)

      expect(wrapper).toMatchSnapshot()
    })
  })
})
