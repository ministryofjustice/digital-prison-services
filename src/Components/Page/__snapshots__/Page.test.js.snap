// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Page /> when there is an error should display spinner and contents 1`] = `
<Fragment>
  <Spinner />
  <Component
    homeLink="http://home/"
  />
  <BackLink
    href="http://this/"
  >
    Back
  </BackLink>
  <styled.div>
    <styled.div>
      <H1
        size="LARGE"
      >
        Page title
      </H1>
    </styled.div>
    <div
      className="page-content"
    >
      Page content
    </div>
  </styled.div>
</Fragment>
`;
